hetio/matrix.py

```diff
@@ -92,34 +69,21 @@ def auto_convert(matrix, threshold):
     Parameters
     ==========
     matrix : numpy.ndarray or scipy.sparse
-    threshold : float (0 < threshold < 1)
-        percent nonzero above which the matrix is converted to dense
+    dense_threshold : float (0 ≤ dense_threshold ≤ 1)
+        minimum proportion of nonzero values at which to output a dense matrix.
+        Setting to 0 ensures output is dense. Setting to 1 ensures output is
+        sparse, unless matrix has no zero entries (use dense_threshold > 1) to
+        gaurantee sparse output.
 
     Returns
     =======
     matrix : numpy.ndarray or scipy.sparse
     """
-    above_thresh = (matrix != 0).sum() / numpy.prod(matrix.shape) >= threshold
-    if sparse.issparse(matrix) and above_thresh:
+    density = (matrix != 0).sum() / numpy.prod(matrix.shape)
+    densify = density >= dense_threshold
+    sparse_input = scipy.sparse.issparse(matrix)
+    if sparse_input and densify:
         return matrix.toarray()
-    elif not above_thresh:
-        return sparse.csc_matrix(matrix)
-    return matrix
-
-
-def copy_array(matrix, copy=True):
-    """Returns a newly allocated array if copy is True"""
-    assert matrix.ndim == 2
-    assert matrix.dtype != 'O'  # Ensures no empty row
-    if not sparse.issparse(matrix):
-        assert numpy.isfinite(matrix).all()  # Checks NaN and Inf
-    try:
-        matrix[0, 0]  # Checks that there is a value in the matrix
-    except IndexError:
-        raise AssertionError("Array may have empty rows")
-
-    mat_type = type(matrix)
-    if mat_type == numpy.ndarray:
-        mat_type = numpy.array
-    matrix = mat_type(matrix, dtype=numpy.float64, copy=copy)
+    if not sparse_input and not densify:
```

Comment:
Should this be `elif` or is another `if` desired for clarity?

Reply:
Those two constructions are equivalent here because of the return statement in the first if. i.e. if first is true, second if statement will never be considered.